### PR TITLE
Remove defer/stream support from subscriptions

### PR DESCRIPTION
**Pull request by robrichard**
Discussion topic: https://github.com/robrichard/defer-stream-wg/discussions/55

This implements what was agreed on in the November 2022 (Secondary, EU) WG. There was consensus to move forward with disabling defer/stream on subscriptions via a field error. This also adds a new validation rule that prevents @defer or @stream from being used inside a subscription operation without an if argument that can be used to disable defer/stream.

### CODE REVIEW

File: src/execution/execute.ts [hunk 1]
```diff
@@ -1478,90 +1483,15 @@ export const defaultFieldResolver: GraphQLFieldResolver<unknown, unknown> =
  *
  * This function does not support incremental delivery (`@defer` and `@stream`).
  * If an operation which would defer or stream data is executed with this
- * function, each `InitialIncrementalExecutionResult` and
- * `SubsequentIncrementalExecutionResult` in the result stream will be replaced
- * with an `ExecutionResult` with a single error stating that defer/stream is
- * not supported.  Use `experimentalSubscribeIncrementally` if you want to
- * support incremental delivery.
+ * function, the result stream will be replaced with an `ExecutionResult`
+ * with a single error stating that defer/stream is not supported.
```
Comment (yaacovCR) [inline src/execution/execute.ts:1487]: I think that with the second commit, this needs to be tweaked a bit => as far as I can tell, field errors are thrown for each payload attempting to utilize incremental delivery, and so the returned `ExecutionResult` may contain data and potentially multiple errors, one of which being the aforementioned field error.

**Review by yaacovCR — COMMENTED** [review]
Other than small comment above regarding comment, the implementation looks good to me.

More generally, I am also fine with leaving incremental delivery support with subscriptions for a potential later stage to help advance the initial release.

File: src/execution/execute.ts (hunk 1 above)
Comment (robrichard) [reply src/execution/execute.ts]: Thanks for catching. This comment is updated now.

**Review by yaacovCR — APPROVED** [review]

**Comment by robrichard** [comment]
@IvanGoncharov this is ready for review, now implementing what was discussed at the November 2022 (Secondary, EU) WG. The [spec PR](https://github.com/graphql/graphql-spec/pull/742) has also been updated to reflect this.

File: src/validation/rules/DeferStreamDirectiveOnValidOperationsRule.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      if (operation.operation === OperationTypeNode.SUBSCRIPTION) {
+        for (const fragment of context.getRecursivelyReferencedFragments(
+          operation,
+        )) {
+          fragmentsUsedOnSubscriptions.add(fragment.name.value);
+        }
+      }
+    },
+    Directive(node, _key, _parent, _path, ancestors) {
+      const definitionNode = ancestors[2];
```
Comment (robrichard) [inline src/validation/rules/DeferStreamDirectiveOnValidOperationsRule.ts:51]: Is this the correct way to get the top level ancestor (either fragment definition or operation type)?

Comment (IvanGoncharov) [reply src/validation/rules/DeferStreamDirectiveOnValidOperationsRule.ts:51]: Yes, this rule can be simplify but this could be done is subsequent PRs

**Pull request merged**